preview-env/create/action.yml

```diff
@@ -105,16 +120,102 @@ runs:
       ARGOCD_AUTH_TOKEN: ${{ inputs.argocd_token }}
       ARGOCD_OPTS: "--grpc-web"
       ARGOCD_SERVER: ${{ inputs.argocd_server }}
-  - name: Wait for sync (up to 20 minutes)
+  - name: Wait for sync
+    id: wait_for_sync
     shell: bash
     run: |-
       echo "::group::wait for sync"
-      argocd app wait ${{ inputs.app_name }} --timeout 1200 --health
+
+      TIMEOUT=${{ inputs.argocd_wait_for_sync_timeout }}
+      INTERVAL=30   # Interval between retries in seconds
+      ELAPSED=0
+      while [ $ELAPSED -lt $TIMEOUT ]; do
+        ARGOCD_APP_STATUS=$(argocd app get "${{ inputs.app_name }}" -o json | jq -r '.status.health.status')
+        echo "Current status: $ARGOCD_APP_STATUS"
+
+        if [ "$ARGOCD_APP_STATUS" == "Healthy" ]; then
+          echo "App is healthy."
+          break
+        fi
+
+        echo "App is not healthy yet. Retrying in $INTERVAL seconds..."
+        sleep $INTERVAL
+        ELAPSED=$((ELAPSED + INTERVAL))
+        echo "Elapsed time is $ELAPSED seconds."
+      done
+
       echo "::endgroup::"
+
+      if [ "$ARGOCD_APP_STATUS" == "Healthy" ]; then
+        echo "Application reached healthy state."
+        exit 0
+      else
+        echo "Timeout reached. The application did not become healthy. Collecting troubleshooting information."
+        ARGOCD_APP_DETAILS=$(cat <<EOF
+      ArgoCD App details
+      ------------------
+      App health: $(argocd app get ${{ inputs.app_name }} -o json | jq -r '.status.health.status')
+      App sync: $(argocd app get ${{ inputs.app_name }} -o json | jq -r '.status.sync.status')
+      App conditions: $(argocd app get ${{ inputs.app_name }} -o json | jq -r '.status.conditions')
+      State of the latest operation phase: $(argocd app get ${{ inputs.app_name }} -o json | jq -r '.status.operationState.phase')
+      Message from the latest operation phase: $(argocd app get ${{ inputs.app_name }} -o json | jq -r '.status.operationState.message')
+      Sync results:
+      $(argocd app get ${{ inputs.app_name }} -o json | jq '.status.operationState.syncResult[]')
+      EOF
+        )
+        echo "argocd_app_details<<EOF" >> $GITHUB_OUTPUT
+        echo "$ARGOCD_APP_DETAILS" >> $GITHUB_OUTPUT
+        echo "EOF" >> $GITHUB_OUTPUT
+        # ensure output is flushed before exiting
+        # sync
+        exit 1
+      fi
+
     env:
       ARGOCD_AUTH_TOKEN: ${{ inputs.argocd_token }}
       ARGOCD_OPTS: "--grpc-web"
       ARGOCD_SERVER: ${{ inputs.argocd_server  }}
+
+    #########################################################################
+    # Get additional troubleshooting information from Kubernetes events
+    # in case of the app didn't reach healthy state.
+    # Events only get requested if a GitHub token with appropriate rights
+    # is passed from the caller workflow.
+  - name: Setup Teleport
+    if: failure() && inputs.github_token != ''
+    id: setup_teleport
+    uses: teleport-actions/setup@v1
+    with:
+      version: 17.2.3
```

Comment:
We should try to avoid hard-coding this version and instead make it discoverable by Renovate, so we retrieve auto-upgrades.